### ISSUE TITLE
CMAKE_EXPORT_COMPILE_COMMANDS Parser fails if compiler executable has a MSDOS "Short path" name

### ISSUE THREAD

**Issue by nolange** [15knots/cmake4eclipse#31]
On Windows the compile_commands.json faisl to process a few entries, if those are using the MSDOS "Short Path" Variant.

In the json file, some entries use `C:\\PROGRA~2\\Atmel\\AVR8-G~1\\bin\\avr-gcc.exe`, others use `C:\\PROGRA~2\\Atmel\\AVR8-G~1\\bin\\AVR-G_~1.EXE`. I believe it depends whether the soures are generated or not, but I am not sure. (Filesystem is NTFS, so generally short paths should not be needed at all)

```
eclipse.buildId=4.6.3.M20170222-0400
java.version=1.8.0_121
java.vendor=Oracle Corporation
BootLoader constants: OS=win32, ARCH=x86_64, WS=win32, NL=en_US
Framework arguments:  -product org.eclipse.epp.package.cpp.product
Command-line arguments:  -os win32 -ws win32 -arch x86_64 -product org.eclipse.epp.package.cpp.product

de.marw.cmake
Warning
Wed Apr 12 13:57:03 CEST 2017
/uit2/build/Debug/compile_commands.json: No parser for command 'C:\PROGRA~2\Atmel\AVR8-G~1\bin\AVR-G_~1.EXE  -DF_CPU=8000000  -g   -fpack-struct -fshort-enums -ffunction-sections -fdata-sections -Wall -Wpointer-arith -Wsign-compare -fstrict-overflow -Wdeprecated --pipe -fmessage-length=0 -mmcu=attiny84 -Wnoexcept -fno-exceptions -fno-rtti -std=gnu++11 -o CMakeFiles\print_compdefs.dir\_lib\print_compdefs.tmp.cpp.obj -c D:\temp\CDH-821\build\Debug\_lib\print_compdefs.tmp.cpp', skipped

```

**Comment by 15knots** [comment]
`compile_commands.json ` is generated by cmake, it is unlikely to depend on whether your sources are generated or not. 
Do your CMakelists.txt files contain commands that explicitly set the C-compiler to use (like `set( CMAKE_C_COMPILER ...)`, see [cmake toolchains](https://cmake.org/cmake/help/v3.6/manual/cmake-toolchains.7.html#cross-compiling))?

1. As for `C:\\PROGRA~2\\Atmel\\AVR8-G~1\\bin\\avr-gcc.exe`, it should be recognized, since there is a detection regex pattern that matches anything that ends with `-gcc.exe`. 
2. But for the compiler shortly named `AVR-G_~1.EXE`, I see very limited chances doe write a regex that detects  `AVR-G_~1.EXE` as the name of a posix-compliant compiler.
Try to fix your setup in a way that cmake does not generate the short name.  Or simply disable the `CMAKE_EXPORT_COMPILE_COMMANDS Parser` in your project settings, your project seems to not use any `-D`, `-I` or `-U` options for the compiler.

**Comment by nolange** [comment]
I set the Compiler via -DCMAKE_C_COMPILER and similar parameters (full path, not the short names). I am not sure what you are implying here, but the single reason I am using CMake is for cross-compiling, so if you meant that you only want to support the system compiler then thats a big letdown.

I`d want to say that you should just convert to long paths, but that seems tricky in Java, and I would hope eclipse has such a function in the core library. The only other solution I found is (the opposite direction, and using external processes): http://stackoverflow.com/questions/18893284/how-to-get-short-filenames-in-windows-using-java

I really doubt I can convince CMake to use long paths, seems this is there because of legacy reasons.

**Comment by 15knots** [comment]
The CMAKE_EXPORT_COMPILE_COMMANDS Parser has nothing to do with the build itself, it is just there to tell the CDT indexer about pre-processor macros and include directories that are specified on the compiler command line. The C-Editor uses that information to gray-out disabled code and to find include files in non-standard directories, if requested. (Your project does not need it at all.)

The CMAKE_EXPORT_COMPILE_COMMANDS Parser just read the file `compile_commands.json` in the build directory, which is written by **cmake**. The parser has to decide, whether the compiler has posix-compliant command-line option for pre-processor macros and include directories or some different. It does so by extracting the name of the compile executable file name. (You can find the relevant pattern matching rules in [CompileCommandsJsonParser.java](https://github.com/15knots/cmake4eclipse/blob/master/de.marw.cmake/src/main/java/de/marw/cmake/cdt/language/settings/providers/CompileCommandsJsonParser.java#L109). A compiler named `AVR-G_~1.EXE`, as in your case, is definitely not something to put on the matching rules list, I think.)

**Comment by nolange** [comment]
First, AVR-G_~1.EXE is the "short path" to avr-g++.exe. I am pretty sure if you use a longer prefix like arm-none-eabi-gcc.exe then things wont get better.

Secondly, you should not read to much into the simple example I attached - it wasnt intended for more than to illustrate the issue. I know about the Providers, and I have larger projects where this gets alot more interesting as lacking defines will mean the indexer wont find the correct symbols and whole directories will light up in red with missing symbols.

the Parsers job should be exactly to extract the definitions and include paths from CMake (which depend on CMake Variables), otherwise I could just enter them manually.
This aint really funny if you have multiple projects, and have to support 2 compilers and 2 platforms.

A Indexer should do a good job matching the build, if you just compile for the host system you probably arent aware of how difficult this is to do, unless the build system and the IDE can communicate somehow.

In other words, saying I should remove the CC Parser or enter the defintions manually just takes away any value of the parser.

**Comment by 15knots** [comment]
>  to extract the definitions and include paths from CMake...

For me it's unclear what you suggest here. Do you think of a system that parses the `CMakelists.txt` files?

**Comment by nolange** [comment]
> For me it's unclear what you suggest here. Do you think of a system that parses the CMakelists.txt files?
No, one that extracts the **information**, but using the compile_commands.json file is a much better idea than the CMakelists.txt files.

I don`t know where the misunderstandings is coming from, I will try again.

### Usage:
I have several CMake Projects, some are simple, others quite complex. I want eclipse`s indexer to have useful information about the defined macros and include paths - entering all that manually is not a good option.
Thats where the CMAKE_EXPORT_COMPILE_COMMANDS parser should help.

Compilers are gcc`s, in the example above avr-gcc and avr-g++, they are configured with `-DCMAKE_C_COMPILER=avr-gcc -DCMAKE_CXX_COMPILER=avr-g++`

### Problem:
At work we have developers working on Windows, and there the `compile_commands.json` contains short paths like `AVR-G_~1.EXE` (this is `avr-g++`, as set before). I have no way of forcing CMake to avaoid using short paths, end result is that there are big issues with missing definitions in the indexer (read: its practically useless, as the avr library heavily depends on macros to support the dozens of different controllers)

### Supposed solution:
On Windows, you should somehow **convert short paths to long paths** before matching the regex. I have no Windows at home, I will try some things at work.

**Comment by nolange** [comment]
ShowPath.java
`import java.nio.file.Paths;
import java.nio.file.Path;
import java.io.File;
import java.io.IOException;


public class ShowPath {

    public static void main(String[] args) {

	for (String path : args) {
    		System.out.println("Argument:  " + path);

		Path p = Paths.get(path);
		System.out.println("Path fn:   " + p.getFileName().toString());

		File f = new File(path);
		System.out.println("File name: " +f.getName());

		try {
			p = p.toRealPath();

			System.out.println("Realp fn:  " + p.getFileName().toString());
		} catch(IOException e) {}

	}
    }

}`

Output:
`D:\temp>java ShowPath C:\PROGRA~2\GNUTOO~1\62017-~1\bin\AR10B2~1.EXE C:\\PROGRA~2\\GNUTOO~1\\62017-~1\\bin\\AR10B2~1.EXE
Argument:  C:\PROGRA~2\GNUTOO~1\62017-~1\bin\AR10B2~1.EXE
Path fn:   AR10B2~1.EXE
File name: AR10B2~1.EXE
Realp fn:  arm-none-eabi-g++.exe
Argument:  C:\\PROGRA~2\\GNUTOO~1\\62017-~1\\bin\\AR10B2~1.EXE
Path fn:   AR10B2~1.EXE
File name: AR10B2~1.EXE
Realp fn:  arm-none-eabi-g++.exe`

As you see, this would also fix #34. you would just need to call `Paths.get(path_from_jsonfile).toRealPath().getFileName().toString()` on every path (potentially just on Windows).

**Comment by 15knots** [comment]
> Problem:
At work we have developers working on Windows, and there the compile_commands.json contains short paths like AVR-G_~1.EXE (this is avr-g++, as set before). I have no way of forcing CMake to avaoid using short paths, end result is that there are big issues with missing definitions in the indexer (read: its practically useless, as the avr library heavily depends on macros to support the dozens of different controllers)

`avr-g++.exe` will be properly recognized as a C++ compiler with posix-compliant options (-I, -D, -U), `AVR-G_~1.EXE`will not be recognized.
I checked the MinGW makefile generator and cmake does not generate short file names (on windows 7).

Of course, I am aware of your problem: In your setup, cmake generates **short file names**, even if you specify **long names** (`-DCMAKE_CXX_COMPILER=avr-g++`). A google search did not return any results that address the problem, so I would tend to deem your setup being the primary thing that should be fixed.

Nevertheless, you provided ideas to fix your special issue, but I suggest to add a separate `CMAKE_EXPORT_COMPILE_COMMANDS Parser for short file names` instead of modifying the existing one for these reasons:

1. The parser has to be fast! Performing time-consuming file system operations (`Path.toRealPath()`) on each source file name will negatively affect **each** user of the existing parser.
1. The plugin\`s runtime environment is targeted as JRE 1.6, which has no class `java.nio.file.Paths`.
1. Testing will be impossible. The setup that causes cmake to generate short file names, even if it is given a normal file name, is only known by employees of your company.
1. No frills. The parser should just do what the documentation says: Extract Info from `compile_commands.json`. None of the project contributors will be able answer questions or handle issues that arise from code added to resolve your issue.
1. This plugin did never aim to solve [every use case](https://github.com/15knots/cmake4eclipse#why-cmake4eclipse), in times, where CDT itself aims to [support cmake ](https://bugs.eclipse.org/bugs/show_bug.cgi?id=174176). 

Finally, to solve your unique  use case, I would suggest you provide a new plugin to your developers that handles your company-specific use-case. Create a new plugin based on [CompileCommandsJsonParser.java](https://github.com/15knots/cmake4eclipse/blob/master/de.marw.cmake/src/main/java/de/marw/cmake/cdt/language/settings/providers/CompileCommandsJsonParser.java). You want to targeted it for a JRE which has class `java.nio.file.Paths`.  Either publish the sources on github or keep them company-intern (it's EPL 1.0 licensed).. PM me if you need consultancy.

**Comment by nolange** [comment]
I narrowed it down, CMake will use short paths if it encounters a space, and I am not the first one encountering [this issue](https://gitlab.kitware.com/cmake/cmake/issues/16138).
Look, you don' t owe me anything, I am still taking my time to look at this issue and proposing solutions. Means I care about your plugin and I am grateful for it.
Whatever you are doing with this report, I am fine with it. But I`d like to drop the notation that I am doing something that was not intended and no one else will run into it.

Some entry in the README about its limitations would be an option, and I still think a workaround for some existing CMake problems can atleast be discussed. Even if CMake fixes this behavior in a new version, you are apparently supporting existing ones too.
Whatever you do to make this issue known or resolved, won`t just benefit me but everyone using your plugin. You spent your time coding it, and I am spending time trying to help making it better.

Now on to your points:

1.  So far I was only proposing running `Path.toRealPath()` just on the programs, in a typical project there won`t be more than 4 different paths of those (C/C++/ASM/Linker).
    If you cache the resulting ParserLookupResult in a Hashmap of the raw (possibly short-) path, that means at most 4 calls to `Path.toRealPath()` per project. A Hashmap lookup is blazingly fast compared to running a path through all the regex matchers aswell (which I believe is done now for each line) - so if speed is a concern you should use one anyway.
    My guess is that you won`t notice a difference if `Path.toRealPath()` is called 4 times after CMake was run to create the compile_commands.json file.
   
2.  Ok, thats your decision and I respect that.

3.  That's a very wild guess from your side, there are people who wont take the time to resolve issues (or dont even realize the parser has issues).
    Other companies might eat babies like we do, or less mean people could install their toolchain at a path with spaces (which **is the default in Windows**).
   
4.  That makes little sense - it would still do the job it was intended.

5.  Nope, but it has a parser for compile_commands.json files. And this is a valid bug/deficiency I added in a Tab called "Issues".

**Comment by nolange** [comment]
Just tested, and `java.io.File.getCanonicalPath()` would work too, so point 2 can be resolved aswell

**Comment by 15knots** [comment]
@nolange: Thanks for pointing me at the cmake issue. I did some googling myself to find s.th like that, but didn\`t find anything (google filter-bubble?). 
So it turns out to not an exotic setup that causes the problem.

But I still do not like to make the workaround the default. I would tend to either 

1. add a separate parser that has the workaround to that [cmake bug](https://gitlab.kitware.com/cmake/cmake/issues/16138). or
2. add an option to enable the workaround in the property page of the existing parser.

In any case, I want to give Users of this plugin a choice to decide whether they need the workaround or not. (They may use linux, OS X; they may use a version of cmake that fixed the short path issue...)


> So far I was proposing running Path.toRealPath() just on the programs, in a typical project there won't be more than 4 different paths of those (C/C++/ASM/Linker).

Actually, it is only the two compilers. But in the **worst** case, it is run for every source file, which sums up in a none-trivial project. (And yes, there is some caching in the parser!)

**Comment by nolange** [comment]
I would very much prefer version 2, and I guess everyone running into this issue will rather have an option than an otherwise identical Parser.
Another idea would be to make an option to disable compiler detection and manually define the compiler type/syntax. eg. a `Automatic` / `GNU (-Compatible)` dropbox. With `Automatic` beeing the current behaviour - trying to extract the information from filename, the other option just parsing the commandline assuming gcc/g++ syntax - choosing between C/C++ solely on input file extension.

> Actually, it is only the two compilers. But in the worst case, it is run for every source file, which sums up in a none-trivial project. (And yes, there is some caching in the parser!)

If you separate the compiler path from the commandline, look up and cache **just this path**, then you will run the getCanonicalPath Function just once for every different tool.

Im glad we are getting somewhere

**Comment by 15knots** [comment]
> Another idea would be to make an option to disable compiler detection and manually define the compiler type/syntax. eg. a Automatic / GNU (-Compatible) dropbox.

Concerning GNU (-Compatible) compilers, you can enable the 'CDT GCC Build Output Parser' and specify the short path in the 'Compiler command pattern'.

**Comment by nolange** [comment]
> Concerning GNU (-Compatible) compilers, you can enable the 'CDT GCC Build Output Parser' and specify the short path in the 'Compiler command pattern'.

That would not fix any issue in your plugin, and I could always just have used the "CDT GCC Build Output Parser". But that parser is a rather bad solution for multiple reasons.

Just to be clear, I don`t want to type in any paths, but just having a way to tell your plugin that the compilers used are GNU (-like). That way the plugin can skip handling the compiler-detection.

**Comment by 15knots** [comment]
Please test https://bintray.com/15knots/p2-zip/download_file?file_path=cmake4eclipse-1.4.7.zip in your environment and close this this, if it fixes that.

**Comment by nolange** [comment]
Hello,

yes the issue is fixed.
Some notes tough, first you can`t reliably test short path expansion (`testDetermineParserForCommandline_MsdosShortNames`) the short paths are like symlink in that thay only correctly work on the same Filesystem. (In other words, the next time I copy gcc in a folder, the short name might be different).
And the "split at first space character" seems unreliable, some day CMake could use a "path with spaces".

Thanks

**Comment by 15knots** [comment]
'testDetermineParserForCommandline_MsdosShortNames' is disabled, it will not work under linux. It was only used to step into the parsing code and look whether the processing worked (without conversion to long file name).

> And the "split at first space character" seems unreliable, some day CMake could use a "path with spaces".

Wasn*t the cmake issue about `Path with space` yields `short file names`? If cmake some day generates it should also not generate short file names, so the original parsing would match the compiler name.
